index.bs

```diff
@@ -507,65 +507,74 @@ subclass <code>0x01</code>,
 </pre>
 </div>
 
-A USB device |device| <dfn data-lt="match a filter">matches a filter</dfn>
-|filter| if the following steps return <code>match</code>:
+A USB device |device| <dfn data-lt="match a device filter">matches a device
+filter</dfn> |filter| if the following steps return <code>match</code>:
 
   1. Let |deviceDesc| be |device|'s <a>device descriptor</a>.
-  1. If <code>|filter|.{{USBDeviceFilter/vendorId}}</code> is present then, if
+  1. If <code>|filter|.{{USBDeviceFilter/vendorId}}</code> is present and
      <code>|deviceDesc|.idVendor</code> does not equal
      <code>|filter|.{{USBDeviceFilter/vendorId}}</code>, return
      <code>mismatch</code>.
-  1. If <code>|filter|.{{USBDeviceFilter/productId}}</code> is present then, if
+  1. If <code>|filter|.{{USBDeviceFilter/productId}}</code> is present and
      <code>|deviceDesc|.idProduct</code> does not equal
      <code>|filter|.{{USBDeviceFilter/productId}}</code>, return
      <code>mismatch</code>.
-  1. If <code>|filter|.{{USBDeviceFilter/classCode}}</code> is present then,
-     for each of |device|'s interface's |interface| run the following steps:
-       1. Let |desc| be |interface|'s <a>interface descriptor</a>.
-       1. If <code>|filter|.{{USBDeviceFilter/classCode}}</code> is present
-          then, if <code>|desc|.bInterfaceClass</code> is not equal to
-          <code>|filter|.{{USBDeviceFilter/classCode}}</code>, continue to
-          the next interface.
-       1. If <code>|filter|.{{USBDeviceFilter/subclassCode}}</code> is present
-          then, if <code>|desc|.bInterfaceSubClass</code> is not equal to
-          <code>|filter|.{{USBDeviceFilter/subclassCode}}</code>, continue to
-          the next interface.
-       1. If <code>|filter|.{{USBDeviceFilter/protocolCode}}</code> is present
-          then, if <code>|desc|.bInterfaceProtocol</code> is not equal to
-          <code>|filter|.{{USBDeviceFilter/protocolCode}}</code>, continue to
-          the next interface.
-       1. Return <code>match</code>.
-  1. If <code>|filter|.{{USBDeviceFilter/classCode}}</code> is present then, if
+  1. If <code>|filter|.{{USBDeviceFilter/serialNumber}}</code> is present then,
+     let |serialNumber| be the <a>string descriptor</a> with index
+     <code>|deviceDesc|.iSerialNumber</code>. If |device| returns an error when
+     requesting |serialNumber| or |serialNumber| is not equal to
+     <code>|filter|.{{USBDeviceFilter/serialNumber}}</code>, return
+     <code>mismatch</code>.
+  1. If <code>|filter|.{{USBDeviceFilter/classCode}}</code> is present and,
+     for any of |device|'s interface's |interface|, |interface| <a>matches
+     the interface filter</a> |filter|, return <code>match</code>.
```

Comment:
A common pattern is to say "if any of these filters doesn't match, return mismatch; otherwise return match." This sequence isn't like that: you finalize a match early for interface matches, even if subsequent steps might mismatch. That deserves a note about why it's the right choice.

If you're trying to treat interfaces and the top-level device symmetrically, it might be worth introducing an abstraction so that you can actually pass them all to the same algorithm.

Reply:
Note added.